lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,425 @@
+					return Ok(Some(ConfirmedTx { tx, block_header, pos, block_height }));
+				} else {
+					// If any previously-confirmed block suddenly is no longer confirmed, we found
+					// an inconsistency and should start over.
+					log_trace!(self.logger, "Inconsistency: Tx {} was unconfirmed during syncing.", txid);
+					return Err(InternalError::Inconsistency);
+				}
+			}
+		}
+		Ok(None)
+	}
+
+	#[maybe_async]
+	fn sync_unconfirmed_transactions(
+		&self, sync_state: &mut SyncState, confirmables: &Vec<&(dyn Confirm + Sync + Send)>,
+	) -> Result<(), InternalError> {
+		// Query the interface for relevant txids and check whether the relevant blocks are still
+		// in the best chain, mark them unconfirmed otherwise. If the transactions have been
+		// reconfirmed in another block, we'll confirm them in the next sync iteration.
+		let relevant_txids = confirmables
+			.iter()
+			.flat_map(|c| c.get_relevant_txids())
+			.collect::<HashSet<(Txid, Option<BlockHash>)>>();
+
+		for (txid, block_hash_opt) in relevant_txids {
+			if let Some(block_hash) = block_hash_opt {
+				let block_status = maybe_await!(self.client.get_block_status(&block_hash))?;
+				if block_status.in_best_chain {
+					// Skip if the block in question is still confirmed.
+					continue;
+				}
+			}
+
+			for c in confirmables {
+				c.transaction_unconfirmed(&txid);
+			}
+
+			sync_state.watched_transactions.insert(txid);
+		}
+
+		Ok(())
+	}
+
+	/// Returns a reference to the underlying esplora client.
+	pub fn client(&self) -> &EsploraClientType {
+		&self.client
+	}
+}
+
+// Represents the current state.
+struct SyncState {
+	// Transactions that were previously processed, but must not be forgotten
+	// yet since they still need to be monitored for confirmation on-chain.
+	watched_transactions: HashSet<Txid>,
+	// Outputs that were previously processed, but must not be forgotten yet as
+	// as we still need to monitor any spends on-chain.
+	watched_outputs: HashSet<WatchedOutput>,
+	// The tip hash observed during our last sync.
+	last_sync_hash: Option<BlockHash>,
+	// Indicates whether we need to resync, e.g., after encountering an error.
+	pending_sync: bool,
+}
+
+impl SyncState {
+	fn new() -> Self {
+		Self {
+			watched_transactions: HashSet::new(),
+			watched_outputs: HashSet::new(),
+			last_sync_hash: None,
+			pending_sync: false,
+		}
+	}
+}
+
+// A queue that is to be filled by `Filter` and drained during the next syncing round.
+struct FilterQueue {
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	transactions: HashSet<Txid>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	outputs: HashSet<WatchedOutput>,
+}
+
+impl FilterQueue {
+	fn new() -> Self {
+		Self {
+			transactions: HashSet::new(),
+			outputs: HashSet::new(),
+		}
+	}
+
+	// Processes the transaction and output queues, returns `true` if new items had been
+	// registered.
```

Comment:
Could clarify they are registered in `SyncState`.

Reply:
Done.